### PR TITLE
[simpl-schema] Basic types

### PR DESCRIPTION
**Pull request by arichter83**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by arichter83** [comment]
Hey! Sorry for the mess with the commits, I didn't have my head on the right spot.

File: types/simpl-schema/simpl-schema-tests.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        regEx: SimpleSchema.RegEx.Id
+    },
+    createdAt: {
+        type: Date,
+        autoValue: () => new Date(),
+    },
+    title: {
+        type: String,
+        label: "Argument",
+        /* TODO, this does not work: "The containing arrow function captures the global value of 'this' which implicitly has type 'any'."
```
Comment (arichter83) [inline types/simpl-schema/simpl-schema-tests.ts]: Help is appreciated here!

Comment (amcasey) [reply types/simpl-schema/simpl-schema-tests.ts]: If you meant for `this` to be a `SchemaDefinition`, I believe you need to use function syntax, rather than arrow syntax: https://www.typescriptlang.org/docs/handbook/functions.html (see the _Type_ section).  However, `SchemaDefinition` does not appear to have a `value` property, so perhaps you meant something else?

File: types/simpl-schema/simpl-schema-tests.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    },
+});
+
+StringSchema.validate({
+    basicString: "Test",
+    limitedString: "pro",
+    regExpString: "id"
+});
+
+    /*
```
Comment (amcasey) [inline types/simpl-schema/simpl-schema-tests.ts]: Did you intend to delete this comment?  Or, perhaps, turn it into further tests?

Comment (arichter83) [reply types/simpl-schema/simpl-schema-tests.ts]: Okay, done both!

File: types/simpl-schema/simpl-schema-tests.ts (hunk 1 above)
Comment (amcasey) [reply types/simpl-schema/simpl-schema-tests.ts]: Sorry, I'm not familiar with the underlying library, so I don't know quite how it's intended to be used.

I could be misreading, but it looks like `custom?: () => any` is on `SchemaDefinition` but the one in the test is actually on the type of `SchemaDefinition.title` (which doesn't seem to be declared?).

Comment (arichter83) [reply types/simpl-schema/simpl-schema-tests.ts]: The parameter is a hash of `SchemaDefinition`s, see here:

```
interface SimpleSchemaStatic {
   new(schema: {[key: string]: SchemaDefinition} | any[], options?: SimpleSchemaOptions): SimpleSchema;
   new(
     schema: {
       [key: string]: SchemaDefinition
         | BooleanConstructor | StringConstructor | NumberConstructor | DateConstructor
         | ArrayConstructor
         | string | RegExp
         | SimpleSchema
     } | any[],
     options?: SimpleSchemaOptions
   ): SimpleSchema;
```
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/31054/commits/250e3e9de236d4c4d74a8c3ce5f08aff00b13f60#diff-21d2fa2cb025853ec8b9fa257855bc15R70

So `SimpleSchema({title: x})` the `x` should be `SchemaDefinition` and it seems to work for all other `SchemaDefinition` files.

The problem is that the callback function passed to `SchemaDefinition.custom` will be executed in a context (bind? ...?) that set's `this.value` among other parameters, which is not part of `SchemaDefinition` or anything. I think I fixed it now:

```
declare CustomValidationContext {
   value: any;
   ...
}
...
        custom?: (this: CustomValidationContext) => undefined | string | SimpleSchemaValidationError;
```

Works for custom: function() { this.value; } not for arrow functions.

File: types/simpl-schema/index.d.ts [hunk 3]
```diff
@@ -4,22 +4,56 @@
 //                 Qkramer <https://github.com/Qkramer>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 
-export class ValidationContext {
-    constructor(ss: any);
+export interface ValidationContext extends SimpleSchemaValidationContextStatic {
+    // constructor(ss: any);
```
Comment (amcasey) [inline types/simpl-schema/index.d.ts]: Delete this?

File: types/simpl-schema/tslint.json [hunk 4]
```diff
@@ -0,0 +1,6 @@
+{
+  "extends": "dtslint/dt.json",
+  "rules": {
+    "object-literal-shorthand": false
```
Comment (amcasey) [inline types/simpl-schema/tslint.json]: Why is this needed?

**Review by amcasey — APPROVED** [review]
Minor comments.

File: types/simpl-schema/index.d.ts (hunk 3 above)
Comment (arichter83) [reply types/simpl-schema/index.d.ts]: Okay.

File: types/simpl-schema/tslint.json (hunk 4 above)
Comment (arichter83) [reply types/simpl-schema/tslint.json]: Because of this discussion: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/31054#discussion_r239460272

If have to use `custom: function() { this.value; }` to make the context available:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d01679afa6d11acd93f6adaa914d2ab74cf52afd/types/simpl-schema/simpl-schema-tests.ts#L42
